text/1011-object-level-security.rst

```diff
@@ -0,0 +1,320 @@
+enforcement of data access policies across all uses of a database.  This is
+important for security and compliance, but would also be tremendously useful
+for backend-less applications that interact with EdgeDB via HTTP (either via
+GraphQL or EdgeQL-over-HTTP).  A secondary motivation is that this will allow
+implementation of `temporal databases <temporal>`_.
+
+
+Requirements
+============
+
+Although there is no direct dependency, this RFC requires
+`RFC 1010 <1001-global-vars.rst>`_ (globals) to be implemented to be practical.
+See `Interaction with globals`_ below for details.
+
+
+Specification
+=============
+
+This RFC proposes to add a new schema item type -- *Access Policy* -- that is
+defined in the context of an object type and specifies the rules of rewriting
+queries that refer to the enclosing object type.
+
+CREATE ACCESS POLICY
+--------------------
+
+Define a new access control policy for a given object type.
+
+Required capabilities: DDL.
+
+Synopsis::
+
+    {CREATE|ALTER} OBJECT TYPE <type-name> "{"
+        CREATE ACCESS POLICY <name>
+            [ WHEN <condition> ]
+            { PERMIT | RESTRICT }
+            { ALL | READ | WRITE | DELETE } [ ... ]
+            USING (<expr>)
+    "}"
+
+An access policy grants the permission to select, insert or delete objects
+that match the relevant policy expression.  The ``READ`` and ``DELETE``
+policies specify a *visibility* check for existing objects: when ``<expr>>``
+returns a false result for an object, such object will be silently skipped
+in any ``SELECT``, ``UPDATE`` or ``DELETE`` expression that scans it.
+A ``WRITE`` policy specifies a *validity* check for new or updated objects and
+affects ``INSERT`` and ``UPDATE`` expressions: if the proposed object violates
+``<expr>``, an error is raised immediately and the query is aborted.
+
+The optional ``<condition>`` expression is evaluated for every object
+affected by the statement and the policy is applied only if the expression
+evaluates to *true*.  It is essentially equivalent to joining ``<condition>``
+with ``<expr>`` with an ``AND`` operator.  The reason for a standalone clause
+is that it makes it easier to separate *when* a policy is applied from
+*how* a policy is applied.  Also, other access policies do not apply to
+``when`` expressions, but they *do* apply to ``using`` expressions.
+
+If at least one access rule is defined for a type, then all elements become
+invisible/immutable by default and must be made visible/mutable by at least
+one ``PERMIT`` rule.
+
+A ``PERMIT`` rule *adds* all objects, for which the check expression
+evaluates to true, to the set of visible objects.  ``PERMIT`` rules are
+combined using the ``OR`` operator.
+
+A ``RESTRICT`` rule restricts the set of visible objects to only objects for
+which the check expression evaluates to true.  ``RESTRICT`` rules are combined
+using the ``AND`` operator and are applied *after* all ``PERMIT`` policies.
```

Comment:
There is a sentence above that clarifies this:

> If at least one access rule is defined for a type, then all elements become invisible/immutable by default and must be made visible/mutable by at least one ``PERMIT`` rule.

`RESTRICT` rules apply _after_ `PERMIT` and enforce exceptions from what is permitted.  This is the same logic that exists in [Postgres RLS](https://www.postgresql.org/docs/current/sql-createpolicy.html#id-1.9.3.75.6.4).

Reply:
Oh, I had misinterpreted `RESTRICT` as doing the *restriction* when the value is true---that is, inverting the meaning of the boolean.

To do some syntax quibbling, then, I think that my incorrect reading here is more natural: `restrict all using (foo)` seems to imply to me that it ought to restrict *when* `foo` holds. I understand that the terminology is made to mirror postgres, but I think using the verb `RESTRICT` instead of the adjective `RESTRICTIVE` makes this misunderstanding easier to have in our setting. I'm not sure if I have a counterproposal but I think it's worth thinking about.

Reply:
The previous RFC iteration had the `RESTRICT TO` spelling, would that be better?

Reply:
> `RESTRICT` rules apply *after* `PERMIT`

I don't really understand this claim, from a semantic point of view. `AND` is commutative.

Reply:
Yeah, I like `RESTRICT TO`

Reply:
> > `RESTRICT` rules apply _after_ `PERMIT`
> I don't really understand this claim, from a semantic point of view. `AND` is commutative.

OK, so here's how OLS works *semantically* (as in mental model):

- you start with an empty set of objects;
- you *add* to the set with `permit` rules;
- you *subtract* from the set with `restrict to` rules

Naturally, implementation-wise `and` is commutatuve in a declarative `filter` statement, but logically subtraction isn't.